torch/utils/checkpoint.py

```diff
@@ -0,0 +1,119 @@
+import torch
+
+
+def detach_variable(inputs):
+    if torch.is_tensor(inputs):
+        inp = inputs.detach()
+        return inp
+    elif isinstance(inputs, tuple):
+        return tuple(detach_variable(v) for v in inputs)
+    else:
+        raise RuntimeError("Unsupported input type: ", type(inputs).__name__)
+
+
+class CheckpointFunction(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        with torch.no_grad():
+            outputs = run_function(*args)
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *args):
+        assert (torch.autograd.is_checkpoint_valid()), \
+            "Checkpointing is not compatible with .grad(), please use .backward() if possible"
+        inputs = ctx.saved_tensors
+        inputs_list = detach_variable(inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs_list)
```

Comment:
the reasoning behind the inputs_list here and not converting it to tensor is that models may have different #inputs, and the way inputs should be passed to the model (their order, etc.) is something that `checkpoint` function won't know about. Rather, the run function that user passes as the input has that information. So, whatever inputs are passed by user, we dispatch them to the `run_function` (whether tensor, tuple etc.) and that run_function should take care of it.

Reply:
Ok, nvm. `saved_tensors` is guaranteed to be a tuple, so it's ok. However, I'd rather avoid calling them `input_list`, because it's a tuple and not a list. Just do `inputs = detach_variable(ctx.saved_tensors)`.